### PR TITLE
Use multiprocessing when sending to Kafka

### PR DESCRIPTION
**Pull request by dadrian**

### CODE REVIEW

**Comment by dadrian** [comment]
This adds multiprocessing to Kafka streams as a stopgap measure.

File: ztag/stream.py [hunk 1]
```diff
@@ -165,6 +166,26 @@ def take(self, pbout):
     def cleanup(self):
         return self.push(noretry=True)
 
+class _KafkaWorkerProcess(multiprocessing.Process):
+
+    def __init__(self, q, topic, main_producer, cert_producer):
```
Comment (zakird) [inline ztag/stream.py]: Can you pass a producer object across processes?

File: ztag/stream.py [hunk 2]
```diff
@@ -176,17 +197,31 @@ def __init__(self, logger=None, destination=None, *args, **kwargs):
         else:
             raise Exception("invalid destination: %s" % destination)
         host = os.environ.get('KAFKA_BOOTSTRAP_HOST', 'localhost:9092')
+        processes = os.environ.get('ZTAG_KAFKA_PROCESSES', 2)
         self.main_producer = KafkaProducer(bootstrap_servers=host)
         self.cert_producer = KafkaProducer(bootstrap_servers=host)
+        self.queue = multiprocessing.Queue()
+        self.processes = list()
+
+        for pnum in xrange(0, int(processes)):
```
Comment (zakird) [inline ztag/stream.py]: This looks like you're sharing a kafka producer across processes? This seems dangerous unless the library explicitly guarantees that this is safe. I would have expected that each process starts its own producer.

File: ztag/stream.py [hunk 3]
```diff
@@ -176,17 +197,31 @@ def __init__(self, logger=None, destination=None, *args, **kwargs):
...
-        for certificate in pbout.certificates:
-            self.cert_producer.send("certificate", certificate)
-        self.main_producer.send(self.topic, pbout.transformed)
+        self.queue.put(pbout)
 
     def cleanup(self):
+        # Trigger end of the queue
+        for p in self.processes:
+            self.queue.put(None)
+        # God forbid Python have sane abstractions
```
Comment (zakird) [inline ztag/stream.py]: This comment adds no information right. Either explicitly describe the problem or remove the comment.

**Comment by zakird** [comment]
It may be worth setting a max size on the queue to Kafka. Otherwise, we could end up just filling all the memory on the machine with a far less efficient data structure than ztee has implemented.

**Comment by zakird** [comment]
I'm not convinced this code is safe as it's currently written. Can you either (1) fix or (2) document that this makes sense as currently written?

File: ztag/stream.py (hunk 1 above)
Comment (dadrian) [inline ztag/stream.py]: Yes, I confirmed this in the documentation. It's apparently more efficient to pass it between processes than to have multiple instances.

File: ztag/stream.py (hunk 2 above)
Comment (dadrian) [inline ztag/stream.py]: > The producer is thread safe and sharing a single producer instance across threads will generally be faster than having multiple instances.

http://kafka-python.readthedocs.io/en/master/apidoc/KafkaProducer.html

Comment (zakird) [inline ztag/stream.py]: That documentation does not state that it's safe across processes.

File: ztag/stream.py (hunk 1 above)
Comment (zakird) [inline ztag/stream.py]: I'm not sure where you've gotten that from. The documentation does not state that and issues on the project explicitly state not to do this. https://github.com/dpkp/kafka-python/issues/570 states "You should not expect to share a KafkaProducer instance across processes; instead you will need to create a new instance for each process."

**Comment by dadrian** [comment]
@zakird Refactored.

**Comment by zakird** [comment]
While this looks good, please see if this actually solves problems before merging. If it doesn't solve the problem, we may not want to keep around this complexity.

**Pull request closed without merging**